content/blog/2023-12-01-task-lifecycle.md

```diff
@@ -0,0 +1,572 @@
+
+```c
+context_switch()
+  switch_to()
+    __switch_to()
+      _switch()
+        do_switch_64
+          std	r6,PACACURRENT(r13)
+```
+
+One interesting consequence of tasks calling `context_switch()` is that the
+previous task is 'suspended'[^2] right where it saves its registers and puts in
+the new task's values. When it is woken up again at some point in the future it
+resumes right where it left off. So when you are reading the `__switch_to()`
+implementation, you are actually looking at two different tasks in the same
+function.
+
+[^2]: Don't forget that the entire concept of tasks is made up by the kernel:
+    from the hardware's point of view we haven't done anything interesting, just
+    changed some registers.
+
+But it gets even weirder: while tasks that put themselves to sleep here wake up
+inside of `_switch()`, new tasks being woken up for the first time start at a
+completely different location! So not only is the task changing, the `_switch()`
+call might not even return back to `__switch_to()`!
+
+
+## Conclusion
+
+And there you have it, everything<sup>_[citation needed]_</sup> you could ever
+need to know when getting started with tasks. Will you need to know this
+specifically? Hard to say. But hopefully it at least provides some useful
+pointers for understanding the execution model of the kernel.
+
+
+## Questions
+
+The following are some questions you might have (read: I had).
+
+### Do userspace tasks run in the kernel?
+
+Yes, 'userspace' tasks run in the kernel. We don't have dedicated tasks to swap
+out to. Thanks to address space quadrants we don't even need to change the
+active memory mapping: upon entry to the kernel we automatically start using the
+PID 0 mapping.
+
+
+### Where does a task get allocated a PID?
+
+Referring to the hardware PID, used for virtual memory translations, this is
+actually a property of the memory mapping struct (`mm_struct`). You can find a
+PID being allocated when a new `mm_struct` is created, which may or may not
+occur depending on the task clone parameters.
+
+
+### How can the fork and exec syscalls be hooked into for arch specific handling?
+
+Fork (and clone) will always invoke `copy_thread()`. The exec call will invoke
+`start_thread()` when loading a binary file. Any other kind of file (script,
+binfmt-misc) will eventually require some form of binary file to load/bootstrap
+it, so `start_thread()` should work for your purposes.
+
+The task context of the calls is fairly predictable: `current` in
+`copy_thread()` refers to the parent because we are still in the middle of
+copying it. For `start_thread()` we have that `current` refers to the task that
+is going to be the new program because it is just configuring itself.
+
+
+### Where do exceptions/interrupts fit in?
+
+When a hardware interrupt triggers it just stops whatever it was doing and dumps
+us at the corresponding exception handler. Our `current()` value still points to
```

Comment:
current()????

Reply:
I mean it's a macro for `get_current()`. But I removed them